### DISCUSSION TITLE
service graph is not showing average time correctly

### DISCUSSION THREAD

**Discussion by maheddin092** [grafana/tempo#4367] [Q&A]
Hey,

I am implementing traces in my application, here the connection type is "messaging" and the application is using Kafka for communication.
In my application, I am consuming from Kafka inserting the message to an internal queue pulling the message in another thread and processing it, and then sending it to another external app that is not instrumented. I am getting the trace details in Grafana, but is the service graph average time calculated by considering only root span time, can any one help me here to fix this.

LABLES: 
consuming:
![image](https://github.com/user-attachments/assets/88f526bc-ba59-4581-b029-755fb4e27a93)

processing from que:
![image](https://github.com/user-attachments/assets/48f476cb-d379-4b7b-85b0-36b94cf26be9)

sending 
![image](https://github.com/user-attachments/assets/69710755-4834-4afa-94b6-6f087e293ec2)

service graph:
![image](https://github.com/user-attachments/assets/11d59f10-3b22-471e-8ce2-8d9e954b92fa)

trace:
![image](https://github.com/user-attachments/assets/f6d7ec41-2447-413e-83e6-81690d3ad31a)

here the average time should be 3490 ms, it is only considering the root span time which is 1 sec.

**Comment by joe-elliott** [comment]
I believe this is just showing a "service graph" of the entry point into your trace. Asynchronous service graph creation is disabled by default but you can turn it on by setting `enable_messaging_system_latency_histogram` to true under the service graph processor block on the metrics generator config map:

https://grafana.com/docs/tempo/latest/configuration/#metrics-generator

We don't use this setting internally. It was PR'ed by the community. I would love to hear how it works for you.

**Reply by maheddin092** [reply]
Hi,

Even after configuring the results are the same.

Configuration:
![image](https://github.com/user-attachments/assets/a963a009-3bb1-4a03-9be8-9888f0f3fb27)

service graph: 
![image](https://github.com/user-attachments/assets/29073c48-857e-4bc6-883f-aafc1289bcec)

trace : 
![image](https://github.com/user-attachments/assets/6d812f2f-8ab5-4540-9e03-7bbb21ca3590)

Is there any link( context link) that has to be done? between the receive and process span (pulling from the Que) or any labeling I am doing wrong.

I will be very thankful for your reply

**Reply by joe-elliott** [reply]
We don't use this feature, but I assume it works. Are the span kinds correcly labeled as Producer and Consumer?

Here's a metrics generator trouble shooting guide:

https://grafana.com/docs/tempo/latest/troubleshooting/metrics-generator/

**Reply by nic-godz** [reply]
I'm not sure I follow...but is the issue about the "service graph" showing incorrect average and the "trace graph" showing correct?

Have you tried prolong the lifetime of the **pCL_OtelTraceRoot** object for entirety of the underlying span A & B objects? Just to make sure how this will be presented in the "service graph".
